llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```diff
@@ -3099,6 +3101,66 @@ bool SPIRVInstructionSelector::wrapIntoSpecConstantOp(
   return Result;
 }
 
+bool SPIRVInstructionSelector::selectDpdCoarse(Register ResVReg,
+                                               const SPIRVType *ResType,
+                                               MachineInstr &I,
+                                               const unsigned DPdOpCode) const {
+  if (!STI.isShader()) {
+    std::string DiagMsg;
+    raw_string_ostream OS(DiagMsg);
+    I.print(OS, true, false, false, false);
+    DiagMsg += " is only supported in shaders.\n";
+    report_fatal_error(DiagMsg.c_str(), false);
+  }
```

Comment:
Should this be a helper? I feel like shader checks are going to be needed for more than this one set of opcodes.

Second this actually a bit more restrictive than just shader. the [spec says ](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpDPdxCoarse)

> This instruction is only valid in the Fragment [Execution Model](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Execution_Model).

Reply:
I couldn't find a reasonable way to determine the Execution Model from here. I've opened the ticket https://github.com/llvm/llvm-project/issues/167562 and will add a TODO here to update the check when we can. I figure we might as well move it to a helper function later when we can update the check to be more correct.

Reply:
> Mostly because I wanted to have som uniformity for error messages.

Ah I see what you mean! Moved the check and diagnostic to a helper.